### DISCUSSION TITLE
Disable mouse

### DISCUSSION THREAD

**Discussion by jacobtomlinson** [Textualize/textual#183] [Q&A] [answered]
Having mouse support in textual is amazing, but for some things I'd love to be able to disable it completely. Is there any way to turn off mouse events?

**Comment by davep** [comment]
Textual doesn't currently have a setting that will do this, but you could roll a widget like this:

```python
class NoMouse( Static ):
    """Mouse killing widget."""

    DEFAULT_CSS = "NoMouse {display: none;}"

    def on_mount( self ):
        self.capture_mouse( True )
```

and compose it into your app and it should (it does in my testing) eat all the mouse vents.

**Reply by davep** [reply]
On second thoughts it could/should likely inherit from `Widget` rather than `Static`.

**Comment by willmcgugan** [comment, marked as answer]
I can think of a few ways of hacking that in. We could also expose that via a simple method on App perhaps. But what is the use-case? Where would you want to disable the mouse?

Easiest way I can think of currently is to catch the mouse events on app. Something like this:

```python
async def on_event(self, event: events.Event) -> None:
    if isinstance(event, MouseEvent):
        event.prevent_default()
```

**Reply by jacobtomlinson** [reply]
The use case is that I don't want folks to be able to change the focus with the mouse, or with tab. I want to be in control of what widget has focus.

**Reply by willmcgugan** [reply]
Ok, but why though? Open to making it easier, but I’d like to understand what problem it solves.

**Reply by jacobtomlinson** [reply]
I think this ask is coming from a maintenance perspective more than anything. Mouse support in console apps like textual has is awesome but unusual. I don't feel like I have a good mental model in my head for how a mouse should behave in the terminal, I also don't expect my users to have an intuitive feeling for what they can click/drag/scroll/etc. For most things I want to build I don't care about mouse support, it's a nice value add if you want buttons and things, but for my projects it's not a high-priority feature. So from my perspective, I'm maintaining an app which has mouse support that I can't turn off, that isn't really needed for the functionality of the app and results in edge case bugs that I struggle to reason about.

I spent a bunch of time today debugging an issue where my focus code wasn't working as expected and it turned out to be because I was clicking on the terminal window to bring it back to the foreground and it was changing the focus. I don't want to spend development time on mouse-related bugs. Ideally, I just wanted vim style bindings where `:` put one widget in focus and `escape` puts another widget in focus. Maybe I want to think about mouse things in the future, but I'd like to opt-out for now and opt-back in later.

**Reply by willmcgugan** [reply]
I would hope that for most apps, mouse support should be pretty intuitive, and not place any additional burden on development. However I can see why you might want to disable the mouse when building a keyboard-oriented app.

The only question is how to expose that. We could have a `disable_mouse` method you could all on mount. Or a classvar on app, so you could do `ENABLE_MOUSE=False`. Do you have any preference?

**Reply by jacobtomlinson** [reply]
Don't get me wrong the mouse support in textual is awesome. It makes it so easy to implement, the burden is very close to zero, but critically not zero. For example, I think the main issue I was running into was that I didn't implement any focus or hover styling to my widgets and then was accidentally changing focus with the mouse leading to problems. If I had appropriately styled the elements I would've seen what was happening. But I haden't even considered having mouse support and so hadn't bothered doing any mouse-related styling.

In terms of implementation I think the classvar makes most sense given that's how you configure other things like bindings.
